### DISCUSSION TITLE
Type check for `device.queue.writeBuffer()`?

### DISCUSSION THREAD

**Discussion by h-a-n-n-e-s** [gpuweb/gpuweb#4363] [Q&A]
Given some .wgsl code
```wgsl
struct integerParameter {
  a: u32,
  b: u32
}
@group(0) @binding(0) var<uniform> para: integerParameter;
```
I created a bug in the .ts file by writing to the buffer
```ts
a = 5;
b = 42;
device.queue.writeBuffer(integerParameterBuffer, 0, new Float32Array([a, b]));
```
which should be `Uint32Array([a, b])` instead.

Currently, this goes unnoticed and the values of `a` and `b` in the .wgsl code are unpredictable.

I wonder if there is a way for WebGPU to notice the inconsistency and throw an error?

**Comment by greggman** [comment]
WebGPU can not check this because it doesn't know what you're going to use the buffer for.

**Comment by h-a-n-n-e-s** [comment]
I see.
And viewed from the Typescript side, I guess TS has no means to infer the u32 type from the .wgsl code? (For example, through the bind group information? Sorry I'm just fantasizing 😁 )

**Comment by greggman** [comment]
You can use any buffer with any wgsl.   There is no association between any particular wgsl and specific buffers (except for usage flags).  There's also no type information associated with buffers.

Also, if you changed your struct to

```
struct mixedParameter {
  a: u32,
  b: f32
}
```

What type would you expect to pass to `writeBuffer`?

If you wanted to add some kind of enforcement you'd need to do that in some higher level library that added restrictions that WebGPU itself does not have.

You might want to look into a library like [webgpu-utils](https://github.com/greggman/webgpu-utils). It provides functions for setting buffer values.

To be TypeScript typesafe though would require a build system that reads the .wgsl at build time and generates typescript types from them.

**Comment by h-a-n-n-e-s** [comment]
Makes sense, thanks a lot!
(Out of curiosity, there is no meaningful way to use `writeBuffer` for the `mixedParameter` struct, right?)

**Comment by greggman** [comment]
I don't know what you mean by 

> there is no meaningful way to use writeBuffer for the mixedParameter struct, right?)

There's plenty of ways to use something like `mixedParameter`. 

Here's one

```
const arrayBuffer = new ArrayBuffer(8);
const a = new Uint32Array(arrayBuffer, 0, 1);
const b = new Float32Array(arrayBuffer, 4, 1);
a[0] = 123;  // set a as u32
b[0] = 456;  // set b as f32
device.queue.writeBuffer(buffer, 0, arrayBuffer);
```

Here's another

```
const asU32 = new Uint32Array(2);
const asF32 = new Float32Array(asU32.buffer);
asU32[0] = 123;  // set a as u32
asF32[1] = 456;  // set b as f32
device.queue.writeBuffer(buffer, 0, asU32);  // passing as asF32 here will also work
```

`writeBuffer` really only takes a pointer to raw memory. (the arraybuffer behind the typedarray view). the fact that you can pass in typedarray views is just a convenience.

You might find [these articles](https://webgpufundamentals.org/webgpu/lessons/webgpu-memory-layout.html) helpful
